### PR TITLE
Byteskip minus1 fixing

### PR DESCRIPTION
**Pull request by tashrifbillah**
1. Fixed byteskip -1 reading issue #70 

2. Added test for both `nrrd` and `nhdr` files with `byte skip: -1`

3. Added test data with `nrrd` gzip and `nhdr` raw files with `byte skip: -1`

### CODE REVIEW

**Comment by tashrifbillah** [comment]
cc: @ihnorton

**Comment by tashrifbillah** [comment]
Let me know what do you think @addisonElliott

File: nrrd/reader.py [hunk 1]
```diff
@@ -339,48 +339,87 @@ def read_data(header, fh=None, filename=None):
...
-        data = np.fromfile(fh, dtype, sep=' ')
-    else:
-        # Handle compressed data now
-        # Construct the decompression object based on encoding
-        if header['encoding'] in ['gzip', 'gz']:
-            decompobj = zlib.decompressobj(zlib.MAX_WBITS | 16)
-        elif header['encoding'] in ['bzip2', 'bz2']:
-            decompobj = bz2.BZ2Decompressor()
+    # Skip the number of lines requested
+    for _ in range(line_skip):
```
Comment (addisonElliott) [inline nrrd/reader.py]: Should we apply the line_skip **regardless** of whether the byte_skip is set. This is currently set to only apply line_skip if byte_skip != -1, but I'm not sure that's correct.

What do you think? When should line_skip be applied?

File: nrrd/reader.py [hunk 2]
```diff
@@ -339,48 +339,87 @@ def read_data(header, fh=None, filename=None):
...
+        fh.readline()
+
+    if byte_skip>=0:
```
Comment (addisonElliott) [inline nrrd/reader.py]: Try to stick with the syntax used in the library. To stay consistent, you would want the if statement to be:
```
if byte_skip >= 0:
```

File: nrrd/reader.py [hunk 3]
```diff
@@ -339,48 +339,87 @@ def read_data(header, fh=None, filename=None):
...
+
```
Comment (addisonElliott) [inline nrrd/reader.py]: Although this code functionally works correctly, I think it can be made much cleaner. You have a large if/else statement with a lot of repeated code. I would stick to the principle of [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

In other words, a cleaner way is to just change what needs to be changed.

File: nrrd/reader.py [hunk 4]
```diff
@@ -339,48 +339,87 @@ def read_data(header, fh=None, filename=None):
...
+    # Skip the number of lines requested
+    for _ in range(line_skip):
+        fh.readline()
+
+    if byte_skip>=0:
+    
+        # If a compression encoding is used, then byte skip AFTER decompressing
+        if header['encoding'] == 'raw':
+            # Skip the requested number of bytes and then parse the data using NumPy
+            fh.seek(byte_skip, os.SEEK_CUR)
```
Comment (addisonElliott) [inline nrrd/reader.py]: For example, this could be done in one line like this:

```suggestion
            if byte_skip == -1:
                fh.seek(-dtype.itemsize * total_data_points, os.SEEK_END)
            else:
                fh.seek(byte_skip if byte_skip != -1 else total_data_points, os.SEEK_CUR)

```

File: nrrd/tests/test_reading.py [hunk 5]
```diff
@@ -65,9 +62,23 @@ def test_read_detached_header_and_data(self):
...
-        self.assertTrue(data.flags['WRITEABLE'])
+        
+    def test_read_detached_header_and_data_with_byteskip_minus1(self):
+        expected_header = self.expected_header
+        expected_header[u'data file'] = os.path.basename(RAW_DATA_FILE_PATH)
+        expected_header[u'byte skip'] = -1
+
+        data, header = nrrd.read(RAW_BYTESKIP_NHDR_FILE_PATH)
+        
+        print('Given')
```
Comment (addisonElliott) [inline nrrd/tests/test_reading.py]: No need for print's in tests.

File: nrrd/tests/test_reading.py [hunk 6]
```diff
@@ -65,9 +62,23 @@ def test_read_detached_header_and_data(self):
...
+        print(expected_header)
+        
+        print('Found')
+        print(header)
+        
+        np.testing.assert_equal(self.expected_header, header)
+        np.testing.assert_equal(data, self.expected_data)
```
Comment (addisonElliott) [inline nrrd/tests/test_reading.py]: I'm not entirely sure how this tests the behavior of the byte_skip = -1. Since it is just seeking backwards to the beginning of the file.

I would change the data_file and add some garbage to the front of it. That way seeking from the front won't work but seeking backwards will! Then have one test using byte_skip = -1, and one test that should fail using no byte_skip.

Comment (addisonElliott) [reply nrrd/tests/test_reading.py]: With this in mind, don't be afraid to add a new NRRD file to test with. Just use some fake/synthesized data that works for your purpose.

File: nrrd/tests/test_reading.py [hunk 7]
```diff
@@ -187,9 +194,6 @@ def test_read_simple_4d_nrrd(self):
         np.testing.assert_equal(data.dtype, np.float64)
         np.testing.assert_equal(data, np.array([[[[0.76903426]]]]))
 
-        # Test that the data read is able to be edited
-        self.assertTrue(data.flags['WRITEABLE'])
```
Comment (addisonElliott) [inline nrrd/tests/test_reading.py]: Why did you remove this?

This assert has a purpose to ensure that the data is writeable. If the data is not writeable, then something is wrong with the code. I had an issue and that's why I added this.

File: nrrd/reader.py (hunk 2 above)
Comment (tashrifbillah) [reply nrrd/reader.py]: Done

File: nrrd/reader.py (hunk 4 above)
Comment (tashrifbillah) [reply nrrd/reader.py]: Done

File: nrrd/reader.py (hunk 1 above)
Comment (addisonElliott) [reply nrrd/reader.py]: Yeah, I agree with you. I looked at the spec and this is what it said:

> When used in combination with byte skip, the line skipping is done before the byte skipping. The meaning of line skip is not affected by the encoding field.

File: nrrd/reader.py [hunk 8]
```diff
@@ -339,22 +339,24 @@ def read_data(header, fh=None, filename=None):
...
-    # If a compression encoding is used, then byte skip AFTER decompressing
-    if header['encoding'] == 'raw':
-        # Skip the requested number of bytes and then parse the data using NumPy
+    # Skip the number of lines requested
+    for _ in range(line_skip):
+        fh.readline()    
+            
+    # Skip the requested number of bytes or seek backward, and then parse the data using NumPy
+    if byte_skip < -1:
+        raise Exception('Invalid byteskip, allowed values are greater than or equal to -1')
```
Comment (addisonElliott) [inline nrrd/reader.py]: Switch to NRRDError rather than Exception
```suggestion
        raise NRRDError('Invalid byteskip, allowed values are greater than or equal to -1')
```

File: nrrd/reader.py [hunk 9]
```diff
@@ -423,9 +425,8 @@ def read(filename, custom_field_map=None):
     """
 
     """Read a NRRD file and return a tuple (data, header)."""
-
     with open(filename, 'rb') as fh:
         header = read_header(fh, custom_field_map)
         data = read_data(header, fh, filename)
 
-        return data, header
+        return data, header
```
Comment (addisonElliott) [inline nrrd/reader.py]: Would remove extra space at the end of this line.

Any whitespace at the end of lines are not necessary

File: nrrd/reader.py [hunk 10]
```diff
@@ -339,22 +339,24 @@ def read_data(header, fh=None, filename=None):
...
+        fh.readline()    
+            
+    # Skip the requested number of bytes or seek backward, and then parse the data using NumPy
+    if byte_skip < -1:
+        raise Exception('Invalid byteskip, allowed values are greater than or equal to -1')
+    elif byte_skip >= 0:
         fh.seek(byte_skip, os.SEEK_CUR)
+    elif byte_skip == -1 and header['encoding'] not in ['gzip', 'gz', 'bzip2', 'bz2']:
+        fh.seek(-dtype.itemsize * total_data_points, os.SEEK_END)
+    else: # The only case left should be: byte_skip == -1 and header['encoding'] == 'gzip'
```
Comment (addisonElliott) [inline nrrd/reader.py]: This is a nitpicky thing, but just to stick with the syntax of the repository (and my personal preference), I would move this line comment to the next line to be on it's own line.

File: nrrd/reader.py [hunk 11]
```diff
@@ -339,22 +339,24 @@ def read_data(header, fh=None, filename=None):
...
+        byte_skip = -dtype.itemsize*total_data_points
```
Comment (addisonElliott) [inline nrrd/reader.py]:
```suggestion
        byte_skip = -dtype.itemsize * total_data_points
```

File: nrrd/reader.py [hunk 12]
```diff
@@ -380,7 +382,7 @@ def read_data(header, fh=None, filename=None):
 
         # Byte skip is applied AFTER the decompression. Skip first x bytes of the decompressed data and parse it using
         # NumPy
-        data = np.fromstring(decompressed_data[byte_skip:], dtype)
+        data = np.frombuffer(decompressed_data[byte_skip:], dtype)
```
Comment (addisonElliott) [inline nrrd/reader.py]: Why was this changed to fromstring to frombuffer?

Remove whitespace at end of line.

Also, shouldn't this have an if/else statement for byte_skip == -1 to do something like the following if byte_skip == -1...
```
decompressed_data[-dtype.itemsize * total_data_points:]
```

File: nrrd/tests/test_reading.py [hunk 13]
```diff
@@ -130,7 +157,7 @@ def test_read_dup_field_error_and_warn(self):
             self.assertTrue("Duplicate header field: 'type'" in str(w[0].message))
 
             self.assertEqual(expected_header, header)
-            nrrd.reader._NRRD_ALLOW_DUPLICATE_FIELD = False
+            nrrd.reader.ALLOW_DUPLICATE_FIELD = False
```
Comment (tashrifbillah) [inline nrrd/tests/test_reading.py]: Shouldn't this be just `ALLOW_DUPLICATE_FIELD`?

Comment (addisonElliott) [reply nrrd/tests/test_reading.py]: Yes, I **think** I fixed this in master, or maybe I fixed a different occurrence.

**Comment by tashrifbillah** [comment]
I have taken care of the previous failures. Sorry, it took me some time to realize that the version installed in my computer was different from the cloned one. This mismatch attributed to all the mysterious missing lines so far.

We should be good now.

File: nrrd/reader.py (hunk 8 above)
Comment (tashrifbillah) [reply nrrd/reader.py]: I was thinking about that, but than I thought shouldn't it be just Exception? But yeah, we can do that.

File: nrrd/reader.py (hunk 10 above)
Comment (tashrifbillah) [reply nrrd/reader.py]: Okay, doing it.

File: nrrd/reader.py (hunk 11 above)
Comment (tashrifbillah) [reply nrrd/reader.py]: Sure

**Comment by addisonElliott** [comment]
> How does it look now? I understand for all compressed data, `data.flags['WRITEABLE']` is `False` for some reason, but don't know why.

Take a look there:
https://github.com/mhe/pynrrd/pull/68

Here is the relevant information I discovered:

> Use fromstring function when dealing with compressed data because it will copy the data from the string into a Numpy array and allow the data to be editable. If you use frombuffer, then it will simply reference the memory. For a binary string, this ends up with the numpy array not being editable.

File: nrrd/reader.py (hunk 8 above)
Comment (addisonElliott) [reply nrrd/reader.py]: I'm open to hearing your reasoning, but my thought process is to use NRRDError for any errors in pynrrd. I'm treating NRRDError as the exceptions for the module.

File: nrrd/reader.py (hunk 12 above)
Comment (tashrifbillah) [reply nrrd/reader.py]: I have already taken care of this condition in the `if` above.

Comment (tashrifbillah) [reply nrrd/reader.py]: The change was a mistake that resulted in failures:
when `np.frombuffer` is used:
 `data.flag['WRITEABLE'] = False`

But when `np.fromstring` is used:
 `data.flag['WRITEABLE'] = True`

The latter is desired in tests.

Comment (addisonElliott) [reply nrrd/reader.py]: Totally missed that even though I commented on that line...yes, you're right. I like the way you're doing it...

**Comment by tashrifbillah** [comment]
Done, sorry about that.

File: nrrd/reader.py (hunk 12 above)
Comment (addisonElliott) [reply nrrd/reader.py]: Yeah, I think we should make the data writeable so we want to use np.fromstring. I had an issue where I loaded gzip NRRD data and tried to edit it but it didn't work because it wasn't writeable.

**Comment by addisonElliott** [comment]
Code looks good, just add a few quick tests to test it and we should be good to go! 👍

**Comment by tashrifbillah** [comment]
> Test that it throws an exception when byte_skip is negative (but not -1)
Change the byte_skip dataset to add fake data to the front of it so that you have to do a byte_skip=-1 to get it to work. Then add a test with/without byte_skip being -1 to ensure it works.
Repeat the above test for raw, ASCII & GZIP

Your test data addition suggestion should be straightforward. I'll try to complete them soon.

**Comment by tashrifbillah** [comment]
> Test that it throws an exception when byte_skip is negative (but not -1)

This is done using [numpy.testing.assert_raises](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.testing.assert_raises.html#numpy.testing.assert_raises). 

But, I don't think you can add fake data in front to test byte_skip -1 functionality because size of the data is fixed by `sizes` property in the header. Adding fake data will raise and `NRRDError` at [Ln 408](https://github.com/mhe/pynrrd/blob/master/nrrd/reader.py).

The purpose of putting byte_skip -1 is to get data from a detached data file even when you have data at the end of an nrrd file.

So, I think I can add a test that reads data from a detached file which is pointed by the header key `data file` in the nrrd file. The above alien data reading is controlled by `byte_skip: -1`.

What do you think?

**Comment by addisonElliott** [comment]
Wasn't aware of numpy.testing.assert_raises so that's great, but the NRRD test util has it's own function, see here:
https://github.com/mhe/pynrrd/blob/master/nrrd/tests/test_reading.py#L122

I would use that just to be inconsistent.

You edited your comment but you make a good point about adding random data. Not sure...but it would be nice to verify that byte_skip=-1 is working correctly in a sense that it won't work without byte_skip = -1. 

If it can't be done then that's okay, just stick with the normal tests.

**Comment by tashrifbillah** [comment]
Yeah, after putting my comment, I realized there might be a way of doing the way you suggested. So, let's see if I can come up with that design. Otherwise, you have my all ears.

**Comment by tashrifbillah** [comment]
Hey @addisonElliott , I have been able to form a test case that demonstrates the use of `byte_skip: -1`. 
I am at the final stage of my editing. However, can you help a little but about https://github.com/mhe/pynrrd/blob/master/nrrd/tests/test_reading.py#L122 ?

I mean how do I apply it instead of
>         np.testing.assert_raises(nrrd.errors.NRRDError, nrrd.read, GZ_NIFTI_NHDR_FILE_PATH)

that I have used?

**Comment by tashrifbillah** [comment]
In my opinion, since you have used `np.testing` all through the `test_reading.py`, I think you are better off using `np.testing.assert_raises`.

**Comment by addisonElliott** [comment]
Yeah, I understand. The benefit of the utility assertRaises function is that you can ensure that the error message is exactly what you expect. Not sure if np.testing has that capability.

Either way works, but for consistency, just stick with the utility function.

It works like this: take that line of code, change the message to what it should be, and then inside the context of the with statement, place your core that should cause an exception

**Comment by addisonElliott** [comment]
Should be something like:
```
with self.assertRaises(nrrd.NRRDError, 'Message here of exception'):
    nrrd.read(GZ_NIFTI_NHDR_FILE_PATH)
```

**Comment by tashrifbillah** [comment]
Can you see the error:

```
    def test_read_detached_header_and_data_with_byteskip_minus5(self):
        
        with self.assertRaises(nrrd.NRRDError, 'Invalid byteskip, allowed values are greater than or equal to -1'):
            nrrd.read(RAW_INVALID_BYTESKIP_NHDR_FILE_PATH)
```

```
ERROR: test_read_detached_header_and_data_with_byteskip_minus5 (test_reading.TestReadingFunctions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tb571/anaconda3/lib/python3.6/site-packages/nrrd/tests/test_reading.py", line 107, in test_read_detached_header_and_data_with_byteskip_minus5
    with self.assertRaises(nrrd.NRRDError, 'Invalid byteskip, allowed values are greater than or equal to -1'):
  File "/home/tb571/anaconda3/lib/python3.6/unittest/case.py", line 733, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/home/tb571/anaconda3/lib/python3.6/unittest/case.py", line 178, in handle
    callable_obj(*args, **kwargs)
TypeError: 'str' object is not callable
```

**Comment by tashrifbillah** [comment]
Never mind, I missed the 'Regex'.

File: nrrd/reader.py [hunk 14]
```diff
@@ -339,22 +339,30 @@ def read_data(header, fh=None, filename=None):
...
+    if line_skip >= 0:
         for _ in range(line_skip):
             fh.readline()
-
-    # If a compression encoding is used, then byte skip AFTER decompressing
-    if header['encoding'] == 'raw':
-        # Skip the requested number of bytes and then parse the data using NumPy
+    else:
+        raise NRRDError('Invalid lineskip, allowed values are greater than or equal to 0')
+
```
Comment (tashrifbillah) [inline nrrd/reader.py]: Also added a sanity check for line skip as noted [here](http://teem.sourceforge.net/nrrd/format.html)

File: nrrd/tests/data/BallBinary30x30x30_byteskip_minus_one_nifti.nhdr [hunk 15]
```diff
@@ -0,0 +1,14 @@
+NRRD0004
+# Complete NRRD file format specification at:
+# http://teem.sourceforge.net/nrrd/format.html
+type: short
+dimension: 3
+space: left-posterior-superior
+sizes: 30 30 30
+byte skip: -1
```
Comment (tashrifbillah) [inline nrrd/tests/data/BallBinary30x30x30_byteskip_minus_one_nifti.nhdr]: The nifti file has the following header at the beginning before accessing image data:

```
sizeof_hdr      : 348
data_type       : b''
db_name         : b''
extents         : 0
session_error   : 0
regular         : b'r'
dim_info        : 0
dim             : [ 3 30 30 30  1  1  1  1]
intent_p1       : 0.0
intent_p2       : 0.0
intent_p3       : 0.0
intent_code     : none
datatype        : int16
bitpix          : 16
slice_start     : 0
pixdim          : [1. 1. 1. 1. 1. 0. 0. 0.]
vox_offset      : 0.0
scl_slope       : nan
scl_inter       : nan
slice_end       : 0
slice_code      : unknown
xyzt_units      : 10
cal_max         : 0.0
cal_min         : 0.0
slice_duration  : 0.0
toffset         : 0.0
glmax           : 0
glmin           : 0
descrip         : b'BYTESKIP -1 TEST'
aux_file        : b''
qform_code      : scanner
sform_code      : scanner
quatern_b       : 0.0
quatern_c       : 0.0
quatern_d       : 1.0
qoffset_x       : 0.0
qoffset_y       : 0.0
qoffset_z       : 0.0
srow_x          : [-1.  0.  0.  0.]
srow_y          : [ 0. -1.  0.  0.]
srow_z          : [0. 0. 1. 0.]
intent_name     : b''
magic           : b'n+1'
```

When we add `byteskip: -1`, the nifti data file as accessed from the end, only the required number of bytes. This way, we are introducing a great way of `nifti` to `nrrd` conversion. We should not need to use `DWIConvert` anymore. Just make a detached header, and point to existing nifti file as the `data file` and we are done. At the PNL, BWH we are developing our current nifti pipeline dealing with nrrd files this way.

File: nrrd/tests/data/BallBinary30x30x30_nifti.nhdr [hunk 16]
```diff
@@ -0,0 +1,14 @@
+NRRD0004
+# Complete NRRD file format specification at:
+# http://teem.sourceforge.net/nrrd/format.html
+type: short
+dimension: 3
+space: left-posterior-superior
+sizes: 30 30 30
+# byte skip: -1
```
Comment (tashrifbillah) [inline nrrd/tests/data/BallBinary30x30x30_nifti.nhdr]: The `byte skip` has been intentionally commented out here to point out the error when nifti file is attempted to read from the beginning.

**Comment by tashrifbillah** [comment]
I hope we have done what we wanted. Now `Pynrrd` supports `byte skip: -1` even for compressed data.

**Comment by addisonElliott** [comment]
Great job @tashrifbillah!

The test cases are great, and I'm especially pleased by the byte_skip=-1 with NIFTI TAR file as the data. Great practical test.

I'm going to merge this into master and I'll wait a few weeks before publishing a new version in case any other PRs are submitted. If I forget in a few weeks, bug me somehow and I'll publish a new version.

👍

**Comment by tashrifbillah** [comment]
No problem, happy to work on it. @ihnorton may also put his two cents in when he gets back.
